cpp/src/parquet/encoding.cc

```diff
@@ -127,6 +129,21 @@ class PlainEncoder : public EncoderImpl, virtual public TypedEncoder<DType> {
   }
 
  protected:
+  template <typename ArrayType>
```

Comment:
i'm not very familiar with this code, could you let me know what these changes are intended to do?

Reply:
`Put(const ::arrow::Array&)` was hardwired for narrow binary arrays, we need to handle both narrow and large arrays, hence this small refactor.